server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderator.java

```diff
@@ -1138,21 +1164,26 @@ public void registerNewVersionOfPendingSegment(
 
     // Announce segments
     final DataSegment baseSegment = sinks.get(basePendingSegment).getSegment();
+    final DataSegment newSegment = getUpgradedSegment(baseSegment, newSegmentVersion);
 
-    final DataSegment newSegment = new DataSegment(
-        newSegmentVersion.getDataSource(),
-        newSegmentVersion.getInterval(),
-        newSegmentVersion.getVersion(),
+    segmentAnnouncer.announceSegment(newSegment);
+    baseSegmentToUpgradedSegments.get(basePendingSegment).add(newSegmentVersion);
+    upgradedSegmentToBaseSegment.put(newSegmentVersion, basePendingSegment);
```

Comment:
These two operations can be put inside a method. Also, I guess it is better to use `computeIfAbsent` for `baseSegmentToUpgradedSegments`.